driver-sync/src/test/functional/com/mongodb/client/RetryableWritesProseTest.java

```diff
@@ -179,6 +188,74 @@ public static <R> void poolClearedExceptionMustBeRetryable(
+            private final AtomicBoolean configureFailPoint = new AtomicBoolean(true);
+
+            @Override
+            public void commandSucceeded(final CommandSucceededEvent event) {
+                if (event.getCommandName().equals("insert") && configureFailPoint.compareAndSet(true, false)) {
+                    Assertions.assertTrue(futureFailPointFromListener.complete(FailPoint.enable(
+                            new BsonDocument()
+                                    .append("configureFailPoint", new BsonString("failCommand"))
+                                    .append("mode", new BsonDocument()
+                                            .append("times", new BsonInt32(1)))
+                                    .append("data", new BsonDocument()
+                                            .append("failCommands", new BsonArray(singletonList(new BsonString("insert"))))
+                                            .append("errorCode", new BsonInt32(10107))
+                                            .append("errorLabels", new BsonArray(Stream.of("RetryableWriteError", "NoWritesPerformed")
+                                                    .map(BsonString::new).collect(Collectors.toList())))),
+                            primaryServerAddress
+                    )));
+                }
+            }
+        };
+        BsonDocument failPointDocument = new BsonDocument()
+                .append("configureFailPoint", new BsonString("failCommand"))
+                .append("mode", new BsonDocument()
+                        .append("times", new BsonInt32(1)))
+                .append("data", new BsonDocument()
+                        .append("writeConcernError", new BsonDocument()
+                                .append("code", new BsonInt32(91))
+                                .append("errorLabels", new BsonArray(Stream.of("RetryableWriteError")
+                                        .map(BsonString::new).collect(Collectors.toList()))))
+                        .append("failCommands", new BsonArray(singletonList(new BsonString("insert")))));
+        try (MongoClient client = clientCreator.apply(getMongoClientSettingsBuilder()
+                .retryWrites(true)
+                .addCommandListener(commandListener)
+                .applyToServerSettings(builder ->
+                        // see `poolClearedExceptionMustBeRetryable` for the explanation
+                        builder.heartbeatFrequency(50, TimeUnit.MILLISECONDS))
+                .build());
+             FailPoint ignored = FailPoint.enable(failPointDocument, Fixture.getPrimary(client))) {
+            MongoCollection<Document> collection = client.getDatabase(getDefaultDatabaseName())
+                    .getCollection("originalErrorMustBePropagatedIfNoWritesPerformed");
+            collection.drop();
+            try {
+                collection.insertOne(new Document());
+            } catch (MongoWriteConcernWithResponseException e) {
```

Comment:
If it is leaking, which I don't think is the case, then `MongoClientImpl` is also leaking in the same way: both classes are `public`, and user may use `catch`/`instanceof` to check for these types. I will replace it with `MongoException` to avoid depending on an unnecessary strict type, given that the type is not part of the public API.

Reply:
That's not quite what I mean.  If I recall correctly, we never intended for _instances_ of this class to be thrown all the way to the application.  Rather, I think the intention was for it to be thrown, caught internally by the driver, and eventually re-thrown as a different exception, e.g. a `MongoBulkWriteException`.    I don't remember the details of why these convolutions were necessary, but it seems to have originated in the commit for JAVA-2847.

Reply:
I created https://jira.mongodb.org/browse/JAVA-4775 to track this bug, which I introduced in 4.4.0 (more details are in the ticket).